### DISCUSSION TITLE
Is there an order to `SEND_COMPLETE` events?

### DISCUSSION THREAD

**Discussion by ManickaP** [microsoft/msquic#3189] [Q&A] [answered]
When NOT using send buffering and when having multiple outstanding sends in the pipeline, it there an order in which `SEND_COMPLETE` events will come, or is it random (based on packet loss etc.)?

I know there's is a context object I can use to identify which complete belongs to which send. I'm asking because from my local tests, the events arrive in order, but I'm pretty sure my environment is not representative of the real world.

**Comment by nibanks** [comment, marked as answer]
They will always come in order with our current design. When we originally wrote this code years back I had the same question: Should we complete sends in the order they are acknowledged or the order they are queued? We ultimately decided to do queue order in case someone migrated from TCP and assumed it had the same behavior.
